### PR TITLE
Support for multi-queries using ;; as a delimiter #21

### LINKED ISSUE #21: Support for CTE, variables, EXECUTE IMMEDIATE
Trying to run queries like this (also tried w/ CTE) fails without a really specific error. The SQL runs perfectly against the raw database, but not within this plugin.

I think part of the problem is the use of `SQL_CALC_FOUND_ROWS` which I've seen inserted during my testing, but would be grateful for any advice on how to get these kinds of queries to run.

```sql
SET @cols = (SELECT GROUP_CONCAT(CONCAT('`', COLUMN_NAME, '`')) FROM information_schema.COLUMNS WHERE TABLE_NAME='test_table_1' AND (COLUMN_NAME NOT LIKE '%Revenue%'));
EXECUTE IMMEDIATE CONCAT('SELECT ', @cols, ' FROM `test_table_1`');
```

### PR DESCRIPTION
**Pull request by dxdc**
Fixes #21

### CODE REVIEW

File: wp-admin-ui/Admin.class.php [hunk 1]
```diff
@@ -2058,6 +2058,12 @@ function get_data( $full = false ) {
 			$wpdb->show_errors( true );
 		}
 
+		$multi_sql = preg_split('/;{2,}\s+/', $sql, -1, PREG_SPLIT_NO_EMPTY);
+		if (count($multi_sql) > 1) {
+			$sql = array_pop($multi_sql);
+            		$multi = mysqli_multi_query($wpdb->dbh, implode('; ', $multi_sql));
+		}
```
Comment (sc0ttkclark) [inline wp-admin-ui/Admin.class.php]:
```suggestion
		// Check for the existence of multiple queries to run separated by ";;".
		$multi_sql = preg_split( '/;{2,}\s+/', $sql, -1, PREG_SPLIT_NO_EMPTY );

		// Check if we have multiple queries to run.
		if ( 1 < count( $multi_sql ) ) {
			// Get the last query and remove it from the array.
			$sql = array_pop( $multi_sql );

			// Execute the other queries altogether.
			$multi = mysqli_multi_query( $wpdb->dbh, implode( '; ', $multi_sql ) );
		}
```

File: wp-admin-ui/Admin.class.php [hunk 2]
```diff
@@ -2058,6 +2058,18 @@ function get_data( $full = false ) {
...
+		// Check for the existence of multiple queries to run separated by ";;".
+		$multi_sql = preg_split( '/;{2,}\s+/', $sql, -1, PREG_SPLIT_NO_EMPTY );
+
+		// Check if we have multiple queries to run.
+		if ( 1 < count( $multi_sql ) ) {
+			// Get the last query and remove it from the array.
+			$sql = array_pop( $multi_sql );
+
+			// Execute the other queries altogether.
+			$multi = mysqli_multi_query( $wpdb->dbh, implode( '; ', $multi_sql ) );
```
Comment (sc0ttkclark) [inline wp-admin-ui/Admin.class.php]: What's the intent of `$multi` here? Will it get used?

Comment (dxdc) [reply wp-admin-ui/Admin.class.php]: It's a boolean response, so potentially could evaluate it somehow if it was false? Not sure if it's easy to integrate to your existing error handling.

**Comment by sc0ttkclark** [comment]
Can you provide an example query set for this that would work as expected?

**Comment by dxdc** [comment]
> Can you provide an example query set for this that would work as expected?

Just modify as needed:

```sql
SET @cols = 'id';;
EXECUTE IMMEDIATE CONCAT(' SELECT ', @cols, ' FROM `my_table_1` LIMIT %%LIMIT%%')
```

**Comment by sc0ttkclark** [comment]
So to confirm -- `EXECUTE IMMEDIATE CONCAT(' SELECT ', @cols, ' FROM `my_table_1` LIMIT %%LIMIT%%')` runs the same as a normal SELECT query runs? Also, based on the logic of this PR, the `SET @cols` runs from the mysqli multi context, and a follow up query on the same connection still gets access to that `@cols` part, correct? Just want to make sure I'm following all of the intention behind this.

**Comment by dxdc** [comment]
@sc0ttkclark -- exactly.

Essentially, everything except the very last line of SQL (using the `;; ` delimiter) will run in the multi query.

Without this PR, the following query works just fine:
(note `EXECUTE IMMEDIATE` requires mariadb >= 10.2.3)

```sql
EXECUTE IMMEDIATE CONCAT(' SELECT ', '*', ' FROM my_table_1 LIMIT %%LIMIT%%')
```

However, this takes away from the power of SQL to use dynamic columns (for example). The variables are retained by session, which I presume is tied to $wpdb and at least stored for the session. I tested it on my end, it works fine.

This is just an example, however. The results from a temporary table could be stored and then pulled out again by a standard SELECT in the last line.

File: wp-admin-ui/Admin.class.php (hunk 2 above)
Comment (sc0ttkclark) [reply wp-admin-ui/Admin.class.php]:
```suggestion
			// Check if we still have multiple queries to run.
			if ( 1 < count( $multi_sql ) ) {
				// Split the queries back up with a regular ";" delimiter.
				$multi_sql = implode( '; ', $multi_sql );

				// Run all queries.
				mysqli_multi_query( $wpdb->dbh, $multi_sql );
			} else {
				// Just get the query.
				$multi_sql = reset( $multi_sql );

				// Run the only query we have.
				$wpdb->query( $multi_sql );
			}
```

**Review by sc0ttkclark — APPROVED** [review]

**Comment by sc0ttkclark** [comment]
Take a look and let me know if the current code fits your needs.

**Comment by dxdc** [comment]
Looks good! Here is a more complex case for you to test on your end (tested on mine against your latest commit).
Just substitute `wp_options` and `wp_database` with the appropriate options table name and database name.

```sql
SET @cols = (SELECT GROUP_CONCAT(CONCAT('`', COLUMN_NAME, '`')) FROM information_schema.COLUMNS WHERE TABLE_NAME='wp_options' AND TABLE_SCHEMA='wp_database' AND (COLUMN_NAME NOT LIKE '%_value%'));;
EXECUTE IMMEDIATE CONCAT('SELECT ', @cols, ' FROM `wp_options` LIMIT %%LIMIT%%');
```

**Comment by sc0ttkclark** [comment]
Sounds good, I'll merge this once I get some docs put some place in the interface on how to use this feature

**Comment by sc0ttkclark** [comment]
I love the idea of this but it opens the door to non-select queries which I'm not entirely sure how best to do in a very secure way. I'm going to close this but I would totally be open to a PR that introduces a filter/action that can allow you to do this same thing for your own custom use through hooks.

**Comment by dxdc** [comment]
@sc0ttkclark thanks for getting all of these merged!

As for the security-related comment, do you mean the fact that the use of a double-semicolon isn't a standard delimiter? Otherwise:

- `$wpdb->get_results` calls `$wpdb->query` already, so I don't see a big difference there.
- `mysqli_multi_query` I can understand a bit more, although there are inherent security risks to adding custom SQL, and I'm not sure that one query vs. several really changes things that much.

What about just adding a setting to enable this feature (at your own risk), even if it's via a wp-config parameter?

Taking another view of this, one of the main reasons I was interested in this feature is to limit the columns shown in certain reports for users with lower permissions. It's possible that something as simple as a regex-based include or exclude would make things simpler. In tables with >50 columns (which are constantly getting modifications), it is helpful to allow some more flexibility in this regard since SQL itself doesn't have great ways of doing dynamic column picking.

Still allowing CTE's has a lot of utility, so I can see the value in having both :) I'm not really sure of the benefit of a custom filter or action since I'm still interested in viewing the results in a given report.

**Pull request closed without merging**